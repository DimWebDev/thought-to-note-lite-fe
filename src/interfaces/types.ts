// Represents a complete note, as returned by the backend
export interface NoteType {
  id: number;
  title: string;
  content: string;
  createdAt: string;
  updatedAt: string;
}

// Represents a new note being created, before it's saved by the backend. 
// updatetAt and createdAt are not needed, since they are generated automatically by the Backend
export interface NewNoteType {
  title: string;
  content: string;
}


export {};