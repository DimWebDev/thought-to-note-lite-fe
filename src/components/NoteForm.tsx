import React, { useState } from 'react';
import { TextField, Button } from '@mui/material';
import styled from '@emotion/styled';

const FormContainer = styled.form`
  display: flex;
  flex-direction: column;
  gap: 1rem;
  margin-bottom: 2rem;
`;

const NoteForm: React.FC<{ addNote: (note: any) => void }> = ({ addNote }) => {
  const [title, setTitle] = useState('');
  const [content, setContent] = useState('');

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    const newNote = {
      id: Date.now(),
      title,
      content,
    };
    addNote(newNote);
    setTitle('');
    setContent('');
  };

  return (
    <FormContainer onSubmit={handleSubmit}>
      <TextField
        label="Title"
        variant="outlined"
        value={title}
        onChange={(e) => setTitle(e.target.value)}
        fullWidth
      />
      <TextField
        label="Content"
        variant="outlined"
        multiline
        rows={4}
        value={content}
        onChange={(e) => setContent(e.target.value)}
        fullWidth
      />
      <Button type="submit" variant="contained" color="primary">
        Save Note
      </Button>
    </FormContainer>
  );
};

export default NoteForm;