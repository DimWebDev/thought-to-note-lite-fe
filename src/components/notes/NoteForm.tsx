import React, { useState } from 'react';
import { TextField, Button } from '@mui/material';
import styled from '@emotion/styled';
import { NewNoteType } from '../../interfaces/types';

const FormContainer = styled.form`
  display: flex;
  flex-direction: column;
  gap: 1rem;
  margin-bottom: 2rem;
`;

interface NoteFormProps {
  addNote: (note: NewNoteType) => void;
}

const NoteForm: React.FC<NoteFormProps> = ({ addNote }) => {
  const [title, setTitle] = useState('');
  const [content, setContent] = useState('');

  const handleSubmit = async (e: React.FormEvent) => {
    e.preventDefault();

    const newNote: NewNoteType = { title, content };

    try {
      addNote(newNote);
      setTitle('');
      setContent('');
    } catch (error) {
      console.error('Error creating note:', error);
    }
  };

  return (
    <FormContainer onSubmit={handleSubmit}>
      <TextField
        label="Title"
        variant="outlined"
        value={title}
        onChange={(e) => setTitle(e.target.value)}
        fullWidth
      />
      <TextField
        label="Content"
        variant="outlined"
        multiline
        rows={4}
        value={content}
        onChange={(e) => setContent(e.target.value)}
        fullWidth
      />
      <Button type="submit" variant="contained" color="primary">
        Save Note
      </Button>
    </FormContainer>
  );
};

export default NoteForm;