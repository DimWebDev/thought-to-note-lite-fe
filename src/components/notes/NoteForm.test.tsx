import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event'; // For realistic user interactions
import NoteForm from './NoteForm';
import { NewNoteType } from '../../interfaces/types';
// '@testing-library/jest-dom' is imported globally in setupTests.ts

describe('NoteForm Component', () => {
  const mockAddNote = jest.fn();

  beforeEach(() => {
    render(<NoteForm addNote={mockAddNote} />);
    jest.clearAllMocks(); // Clear mock history before each test
  });

  it('renders the title and content input fields and the submit button', () => {
    // Check for Title input field
    const titleInput = screen.getByLabelText(/title/i);
    expect(titleInput).toBeInTheDocument();
    expect(titleInput).toHaveAttribute('type', 'text');

    // Check for Content input field
    const contentInput = screen.getByLabelText(/content/i);
    expect(contentInput).toBeInTheDocument();
    // Instead of checking for 'multiline' attribute, verify it's a textarea
    expect(contentInput.tagName).toBe('TEXTAREA'); // tagName is uppercase

    // Check for Save Note button
    const saveButton = screen.getByRole('button', { name: /save note/i });
    expect(saveButton).toBeInTheDocument();
    expect(saveButton).toHaveAttribute('type', 'submit');
  });

  it('allows the user to input a title', async () => {
    const titleInput = screen.getByLabelText(/title/i);
    await userEvent.type(titleInput, 'My New Note');
    expect(titleInput).toHaveValue('My New Note');
  });

  it('allows the user to input content', async () => {
    const contentInput = screen.getByLabelText(/content/i);
    await userEvent.type(contentInput, 'This is the content of my new note.');
    expect(contentInput).toHaveValue('This is the content of my new note.');
  });

  it('calls addNote with correct data and clears the input fields upon submission', async () => {
    const titleInput = screen.getByLabelText(/title/i);
    const contentInput = screen.getByLabelText(/content/i);
    const saveButton = screen.getByRole('button', { name: /save note/i });

    const testTitle = 'Test Note';
    const testContent = 'This is a test note content.';

    await userEvent.type(titleInput, testTitle);
    await userEvent.type(contentInput, testContent);

    await userEvent.click(saveButton);

    expect(mockAddNote).toHaveBeenCalledTimes(1);
    expect(mockAddNote).toHaveBeenCalledWith({
      title: testTitle,
      content: testContent,
    });

    expect(titleInput).toHaveValue('');
    expect(contentInput).toHaveValue('');
  });

  it('calls addNote with empty data when fields are empty and clears the fields', async () => {
    const saveButton = screen.getByRole('button', { name: /save note/i });

    const titleInput = screen.getByLabelText(/title/i);
    const contentInput = screen.getByLabelText(/content/i);
    expect(titleInput).toHaveValue('');
    expect(contentInput).toHaveValue('');

    await userEvent.click(saveButton);

    expect(mockAddNote).toHaveBeenCalledTimes(1);
    expect(mockAddNote).toHaveBeenCalledWith({
      title: '',
      content: '',
    });

    expect(titleInput).toHaveValue('');
    expect(contentInput).toHaveValue('');
  });

  it('handles errors gracefully when addNote throws an error', async () => {
    // Modify the mock to throw an error when called
    mockAddNote.mockImplementation(() => {
      throw new Error('Failed to add note');
    });

    const titleInput = screen.getByLabelText(/title/i);
    const contentInput = screen.getByLabelText(/content/i);
    const saveButton = screen.getByRole('button', { name: /save note/i });

    await userEvent.type(titleInput, 'Error Note');
    await userEvent.type(contentInput, 'This note will fail to save.');

    // Spy on console.error to verify error handling
    const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

    await userEvent.click(saveButton);

    expect(mockAddNote).toHaveBeenCalledWith({
      title: 'Error Note',
      content: 'This note will fail to save.',
    });

    // Assert that the fields are not cleared due to the error
    expect(titleInput).toHaveValue('Error Note');
    expect(contentInput).toHaveValue('This note will fail to save.');

    // Assert that console.error was called with the correct error message
    expect(consoleErrorSpy).toHaveBeenCalledWith('Error creating note:', expect.any(Error));

    // Restore the original console.error implementation
    consoleErrorSpy.mockRestore();
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(<NoteForm addNote={mockAddNote} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
