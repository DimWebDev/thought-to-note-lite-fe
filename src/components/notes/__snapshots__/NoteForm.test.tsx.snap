// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteForm Component matches the snapshot 1`] = `
<DocumentFragment>
  <form
    class="css-1uqcufy"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":re:"
        id=":re:-label"
      >
        Title
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":re:"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Title
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":rf:"
        id=":rf:-label"
      >
        Content
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <textarea
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
          id=":rf:"
          rows="4"
          style="height: 0px; overflow: hidden;"
        />
        <textarea
          aria-hidden="true"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
          readonly=""
          style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
          tabindex="-1"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Content
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="submit"
    >
      Save Note
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
